### PR TITLE
Enabled APK mailing on app generator 

### LINKED ISSUE #1601: Droidgen.eventyay doesn't mail the APK
**Actual Behaviour**
No mail is sent with the generated APK

**Expected Behaviour**
Earlier, it used to send the mail.

### PR DESCRIPTION
**Pull request by mahikaw**
Fixes #1601 

Changes: Enabled app generator(v2) to mail APK via:

- Sendgrid, or
- SMTP 

-----

@niranjan94 @the-dagger @iamareebjamal @Shailesh351 @shreyasnbhat @rohanagarwal94 @arp95 Please review. 

Also, @niranjan94 Could you please specify the environment variables, if any that I should parse using ```envparse``` so that I can make the required changes.

### CODE REVIEW

File: apk-generator/v2/app/utils/notification.py [hunk 1]
```diff
@@ -21,7 +22,46 @@ def send(to, subject, message, file_attachment, via_api):
         :param via_api:
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py:25]: This function is just for choosing which method to use. SMTP related config to be done in the smtp send method. Sendgrid config to be done sendgrid method.

File: apk-generator/v2/app/utils/notification.py [hunk 2]
```diff
@@ -21,7 +22,46 @@ def send(to, subject, message, file_attachment, via_api):
...
+                Notification.send_mail_via_smtp_(config,payload)
+            else:
+                key = current_app.config['SENDGRID_KEY']
+                if not key and not current_app.config['TESTING']:
+                    print ('Sendgrid key not defined')
+                    return
+                headers = {
+                    "Authorization": ("Bearer " + key)
+                }
+                Notification.send_email_task.delay(payload, headers)
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: No `delay` here

File: apk-generator/v2/app/utils/notification.py [hunk 3]
```diff
@@ -46,7 +86,15 @@ def send_mail_via_smtp_(config, payload):
         mailer.start()
         message = Message(author=payload['from'], to=payload['to'])
         message.subject = payload['subject']
-        message.plain = strip_tags(payload['html'])
-        message.rich = payload['html']
+        message.plain = message
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Why are you assigning `message` to a property of `message` ? Should be the message from the payload.

File: apk-generator/v2/config.py [hunk 4]
```diff
@@ -22,6 +22,14 @@ class Config:
     CELERY_BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
     CELERY_RESULT_BACKEND = CELERY_BROKER_URL
     SECRET_KEY = os.environ.get('SECRET_KEY', 'JQgeEYk9b4VjxhJP')
+    EMAIL_SERVICE=os.environ.get('SERVICE_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Space around `=` for all of this.

Comment (niranjan94) [inline apk-generator/v2/config.py]: Call env variable also as `EMAIL_SERVICE` and keep default as sendgrid.

File: apk-generator/v2/config.py [hunk 5]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    SMTP_ENCRYPTION=os.environ.get('ENCRYPTION_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Call env variable also as `SMTP_ENCRYPTION`

File: apk-generator/v2/config.py [hunk 6]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    HOST=os.environ.get('HOST_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Call both variable and env variable as `SMTP_HOST`

File: apk-generator/v2/config.py [hunk 7]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    USERNAME=os.environ.get('USERNAME_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Call both variable and env variable as `SMTP_USERNAME`

File: apk-generator/v2/config.py [hunk 8]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    PASSWORD=os.environ.get('PASSWORD_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Call both variable and env variable as `SMTP_PASSWORD`

File: apk-generator/v2/config.py [hunk 9]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    PORT=os.environ.get('PORT_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Call both variable and env variable as `SMTP_PORT`

File: apk-generator/v2/config.py [hunk 10]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    TESTING=os.environ.get('TESTING_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: What is this ?

File: apk-generator/v2/config.py [hunk 11]
```diff
@@ -22,6 +22,14 @@ class Config:
...
+    SENDGRID_KEY=os.environ.get('SENDGRID_KEY_MAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Call both variable and env variable as `SENDGRID_KEY `

File: apk-generator/v2/app/utils/notification.py [hunk 12]
```diff
@@ -21,7 +22,46 @@ def send(to, subject, message, file_attachment, via_api):
...
+                    'port': current_app.config['PORT'],
+                }
+                Notification.send_mail_via_smtp_(config,payload)
+            else:
+                key = current_app.config['SENDGRID_KEY']
+                if not key and not current_app.config['TESTING']:
+                    print ('Sendgrid key not defined')
+                    return
+                headers = {
+                    "Authorization": ("Bearer " + key)
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Why the brackets ?

**Review by niranjan94 — CHANGES_REQUESTED** [review]
@mahikaw

**Comment by mahikaw** [comment]
Updated @niranjan94 Please review.

File: apk-generator/v2/config.py [hunk 13]
```diff
@@ -22,6 +22,13 @@ class Config:
     CELERY_BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
     CELERY_RESULT_BACKEND = CELERY_BROKER_URL
     SECRET_KEY = os.environ.get('SECRET_KEY', 'JQgeEYk9b4VjxhJP')
+    EMAIL_SERVICE = os.environ.get('EMAIL_SERVICE','sendgrid')
```
Comment (iamareebjamal) [inline apk-generator/v2/config.py]: Add space after comma

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: apk-generator/v2/app/utils/notification.py [hunk 14]
```diff
@@ -21,16 +22,47 @@ def send(to, subject, message, file_attachment, via_api):
         :param via_api:
         :return:
         """
-        pass
+
+        email_service = current_app.config['EMAIL_SERVICE']
+
+        payload = {
+            'to': to,
+            'from': current_app.config['SMTP_USERNAME'],
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: The smtp username may not always be an email address. And why would there be an smtp username if the the user wants to use sendgrid. ? 

So, have a separate variable called as `FROM_EMAIL`

File: apk-generator/v2/app/utils/notification.py [hunk 15]
```diff
@@ -46,7 +78,23 @@ def send_mail_via_smtp_(config, payload):
...
+        message.attach(payload['attachment'], data=None, maintype=None, subtype=None, inline=False)
         mailer.send(message)
         mailer.stop()
+
+    @staticmethod
+    def send_email_task(payload):
+
+        key = current_app.config['SENDGRID_KEY']
+        if not key:
+            print 'Sendgrid key not defined'
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Use the logger to log stuff

File: apk-generator/v2/app/utils/notification.py [hunk 16]
```diff
@@ -46,7 +78,23 @@ def send_mail_via_smtp_(config, payload):
...
         message.subject = payload['subject']
-        message.plain = strip_tags(payload['html'])
-        message.rich = payload['html']
+        message.plain = payload['message']
+        message.attach(payload['attachment'], data=None, maintype=None, subtype=None, inline=False)
         mailer.send(message)
         mailer.stop()
+
+    @staticmethod
+    def send_email_task(payload):
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: call this `send_email_via_sendgrid`

File: apk-generator/v2/app/utils/notification.py [hunk 17]
```diff
@@ -46,7 +78,23 @@ def send_mail_via_smtp_(config, payload):
         mailer.start()
         message = Message(author=payload['from'], to=payload['to'])
         message.subject = payload['subject']
-        message.plain = strip_tags(payload['html'])
-        message.rich = payload['html']
+        message.plain = payload['message']
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: The `message` is a rich text string. (Has html tags). So, have both `message.plain` (with html tags stripped) and `message.rich`

File: apk-generator/v2/app/utils/notification.py [hunk 18]
```diff
@@ -1,12 +1,13 @@
 from marrow.mailer import Mailer, Message
 
-from app.utils import strip_tags
+from flask import current_app
+import requests
+
 
 # TODO Notifications have to be implemented. API/SMTP/Sendgrid depending on a configuration
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Update this comment

**Review by niranjan94 — CHANGES_REQUESTED** [review]
@mahikaw

**Comment by mahikaw** [comment]
@niranjan94 @iamareebjamal , updated it.

File: apk-generator/v2/app/utils/notification.py [hunk 19]
```diff
@@ -1,12 +1,15 @@
 from marrow.mailer import Mailer, Message
 
-from app.utils import strip_tags
+from flask import current_app
+import requests
+from celery.utils.log import get_task_logger
 
-# TODO Notifications have to be implemented. API/SMTP/Sendgrid depending on a configuration
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: I meant ... API implementation is still remaining, so keep the TODO for that

File: apk-generator/v2/app/utils/notification.py [hunk 20]
```diff
@@ -21,16 +24,47 @@ def send(to, subject, message, file_attachment, via_api):
...
+
+        payload = {
+            'to': to,
+            'from': current_app.config['FROM_EMAIL'],
+            'subject': subject,
+            'message': message,
+            'attachment': file_attachment,
+        }
+
+        if email_service == 'smtp':
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Send via smtp or sendgrid, only if `via_api` is `False`

File: apk-generator/v2/app/utils/notification.py [hunk 21]
```diff
@@ -46,7 +80,24 @@ def send_mail_via_smtp_(config, payload):
         mailer.start()
         message = Message(author=payload['from'], to=payload['to'])
         message.subject = payload['subject']
-        message.plain = strip_tags(payload['html'])
+        message.plain = payload['message']
         message.rich = payload['html']
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: There is no param called `html`. Its called as `message`. Put that in rich and use strip tags function on it and use that for plain.

**Review by niranjan94 — CHANGES_REQUESTED** [review]

File: apk-generator/v2/config.py [hunk 22]
```diff
@@ -22,7 +22,14 @@ class Config:
...
     SECRET_KEY = os.environ.get('SECRET_KEY', 'JQgeEYk9b4VjxhJP')
-
+    EMAIL_SERVICE = os.environ.get('EMAIL_SERVICE', 'sendgrid')
+    SMTP_ENCRYPTION = os.environ.get('SMTP_ENCRYPTION')
+    SMTP_HOST = os.environ.get('SMTP_HOST')
+    SMTP_USERNAME = os.environ.get('SMTP_USERNAME')
+    SMTP_PASSWORD = os.environ.get('SMTP_PASSWORD')
+    SMTP_PORT = os.environ.get('SMTP_PORT')
+    SENDGRID_KEY = os.environ.get('SENDGRID_KEY')
+    FROM_EMAIL =os.environ.get('FROM_EMAIL')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Space after `=` please

**Review by niranjan94 — CHANGES_REQUESTED** [review]

File: apk-generator/v2/app/utils/notification.py (hunk 21 above)
Comment (mahikaw) [reply apk-generator/v2/app/utils/notification.py]: um yes. Sorry for not taking care.

File: apk-generator/v2/config.py (hunk 22 above)
Comment (mahikaw) [reply apk-generator/v2/config.py]: Updating 👍

**Comment by mahikaw** [comment]
Updated @niranjan94

File: apk-generator/v2/app/utils/notification.py (hunk 19 above)
Comment (niranjan94) [reply apk-generator/v2/app/utils/notification.py]: Isn't resolved yet @mahikaw

**Comment by mahikaw** [comment]
@iamareebjamal @niranjan94 Please approve if there aren't any further changes required.

File: apk-generator/v2/config.py [hunk 23]
```diff
@@ -22,7 +22,14 @@ class Config:
     CELERY_BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
     CELERY_RESULT_BACKEND = CELERY_BROKER_URL
     SECRET_KEY = os.environ.get('SECRET_KEY', 'JQgeEYk9b4VjxhJP')
-
+    EMAIL_SERVICE = os.environ.get('EMAIL_SERVICE', 'sendgrid')
```
Comment (niranjan94) [inline apk-generator/v2/config.py]: Keep default as `none`. If none, email service should be disabled.

File: apk-generator/v2/app/utils/notification.py [hunk 24]
```diff
@@ -46,7 +81,24 @@ def send_mail_via_smtp_(config, payload):
         mailer.start()
         message = Message(author=payload['from'], to=payload['to'])
         message.subject = payload['subject']
-        message.plain = strip_tags(payload['html'])
-        message.rich = payload['html']
+        message.plain = strip_tags(payload['message'])
+        message.rich = payload['message']
+        message.attach(payload['attachment'], data=None, maintype=None, subtype=None, inline=False)
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Are all the additional `None` arguments required?

File: apk-generator/v2/app/utils/notification.py [hunk 25]
```diff
@@ -21,16 +25,47 @@ def send(to, subject, message, file_attachment, via_api):
...
+        }
+        if not via_api:
+            if email_service == 'smtp':
+                Notification.send_mail_via_smtp_(payload)
+            else:
+                Notification.send_email_via_sendgrid_(payload)
 
     @staticmethod
-    def send_mail_via_smtp_(config, payload):
+    def send_mail_via_smtp_(payload):
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Why the underscore at the end?

File: apk-generator/v2/app/utils/notification.py [hunk 26]
```diff
@@ -46,7 +81,24 @@ def send_mail_via_smtp_(config, payload):
...
-        message.plain = strip_tags(payload['html'])
-        message.rich = payload['html']
+        message.plain = strip_tags(payload['message'])
+        message.rich = payload['message']
+        message.attach(payload['attachment'], data=None, maintype=None, subtype=None, inline=False)
         mailer.send(message)
         mailer.stop()
+
+    @staticmethod
+    def send_email_via_sendgrid_(payload):
```
Comment (niranjan94) [inline apk-generator/v2/app/utils/notification.py]: Why the underscore at the end.?

**Review by niranjan94 — CHANGES_REQUESTED** [review]

File: apk-generator/v2/config.py (hunk 23 above)
Comment (mahikaw) [reply apk-generator/v2/config.py]: In one of the earlier reviews, you suggested to set it to 'send grid'. I'll make it none then and add a condition to check that as well?

File: apk-generator/v2/app/utils/notification.py (hunk 24 above)
Comment (mahikaw) [reply apk-generator/v2/app/utils/notification.py]: I had done this after seeing the Marrow mailer [documentation](https://github.com/marrow/mailer). I think they are required. The maintype can be remove though, I'll do that.

**Comment by niranjan94** [comment]
Yes... I had suggested that. But.. I think the generator itself should work
without the need for an email server. (no emails. Wait and download).

So, have a none option. And keep it default. Also update the Readme file
with details about email.

**Comment by mahikaw** [comment]
@niranjan94 Updated as per suggestions. Please review the changes.

File: apk-generator/v2/app/utils/notification.py (hunk 24 above)
Comment (niranjan94) [reply apk-generator/v2/app/utils/notification.py]: @mahikaw see here https://github.com/marrow/mailer/blob/develop/marrow/mailer/message.py#L254 ... The other arguments are not needed. The defaults are `None` anyway. So remove them please. Once this is done, I think the PR can be merged. 

Good work :)

**Review by niranjan94 — CHANGES_REQUESTED** [review]

**Comment by mahikaw** [comment]
Updated @niranjan94 👍

**Review by niranjan94 — APPROVED** [review]

**Review by rohanagarwal94 — APPROVED** [review]

**Pull request merged**